### ISSUE TITLE
Right way to update User Extended fields from code

### ISSUE THREAD

**Issue by nesjett** [e107inc/e107#3879]
I've been trying to update extended fields from withing the code and It looks really complex, looking into usersettings.php looks like the proper way of doing so, but I think it could be done in just 2 methods, one for sanitizing and the other for actually inserting the data.

Actually, the way I found to do it, based on usersettings.php is like this:

` $d = ['user_gender'=> $genderize->gender];
                $euVals = $ue->sanitizeAll($d);
                //$euVals = $ue->userExtendedValidateAll($euVals, []);		// Validate the extended user fields
                $eufVals['data']['user_gender'] = $tp->toDB($euVals['user_gender']);
                if (!empty($eufVals['data']))
                {
                    $ue->addFieldTypes($eufVals);				// Add in the data types for storage
                    $eufVals['_DUPLICATE_KEY_UPDATE'] = true; // update record if key found, otherwise INSERT.
                    $eufVals['data']['user_extended_id'] = $data['user_id'];

                    if (false === $sql->insert('user_extended', $eufVals))
                    {
                        $log->add('error setting gender', $eufVals, 'E_LOG_INFORMATIVE', 'OAUTH');
                    }

                }`

Althought, this line: 
` //$euVals = $ue->userExtendedValidateAll($euVals, []);		// Validate the extended user fields`
Is used in usersettings.php but I can't get it to work in other places so I had to put results by hand, without other checks with this line:
`$eufVals['data']['user_gender'] = $tp->toDB($euVals['user_gender']);`

Really there is no other way of setting user extended fields?

**Comment by nesjett** [comment]
A detailed look into userExtendedValidateAll() shows that this line is returning false, so the function does not do its job:

`if (1==1 ||($isSignup && (int) $defs['user_extended_struct_applicable'] === (int) e_UC_MEMBER && (int) $defs['user_extended_struct_write'] === (int) e_UC_MEMBER) || (check_class($defs['user_extended_struct_applicable']) && check_class($defs['user_extended_struct_write'])))
				`

I still dont understand it at all, but I guess it's because I'm calling it with the user logged out

**Comment by Moc** [comment]
There are lots of functionalities related to the EUF in core already so I doubt that you'll need to change the code. 

Can you explain the end result that you are trying to achieve?

**Comment by nesjett** [comment]
I just want to modify the value of an user extended field, taking into account its params, for example if its a radio button, it should only accept one of it's configured params.

**Comment by Moc** [comment]
Not sure if I understand you correctly still, but can you not edit that by editing the user in the admin area?

**Comment by nesjett** [comment]
No, I need to edit values programatically, for example, I'm updating the extended user field "gender" when the event user_xup_signup is fired.

This would be perfect:

`e107::user()->setExtendedFront('gender', 'M => UE_LAN_MALE');`

But it checks for the right permissions of the user, this means this cannot be fired when the user is "anonimous"

**Comment by Jimmi08** [comment]
@nesjett sorry, I didn't read it in detail (it's too complicated) but we used the birthday as mandatory field and it worked for anonymous too.

**Comment by nesjett** [comment]
Thank you @Jimmi08 , what I'm asking is just for the best way to modify a extended user field WITHOUT an input element.

I'll check the way signup does it anyway, thank you!

**Comment by Jimmi08** [comment]
I needed this now and this works:

```
 if($result) {
	 $userThreadId = $result['threadid'];
         $ue = new e107_user_extended;
        $ue->user_extended_setvalue($aposUserId, 'user_plugin_apos_forum_thread', $userThreadId)ô
} 
```
